### ISSUE TITLE
[no-unnecessary-type-constraint] false positive for generic type with default type on arrow function in a JSX/TSX file

### ISSUE THREAD

**Issue by AllySummers** [typescript-eslint/typescript-eslint#4062]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**
https://github.com/AllySummers/typescript-eslint-jsx-generics
Repo has eslint v8 and typescript-eslint v5, issue still happens on eslint 7 and typescript-eslint 4 (including alpha).
To install eslint v7/typescript-eslint v4 in that repo do this:
`npm i --save @typescript-eslint/eslint-plugin@4.33.1-alpha.15 @typescript-eslint/eslint-plugin@4.33.1-alpha.15 eslint@7.32.0`

The issue can also be seen from running `npm run lint`

**not-working.tsx**
```ts
export interface GenericComponentProps<T> {
  prop: T
}

export const GenericComponent = <T>(props: GenericComponentProps<T>) => (
  <div>hello</div>
);
// Error: 5:70  error  Parsing error: Unexpected token. Did you mean `{'>'}` or `&gt;`?
```
**working.tsx** (sort-of)
```ts
export interface GenericComponentProps<T> {
  prop: T
}

export const GenericComponent = <T extends unknown = unknown>(props: GenericComponentProps<T>) => (
  <div>hello</div>
);
// Error: 5:34  error  Constraining the generic type `T` to `unknown` does nothing and is unnecessary  @typescript-eslint/no-unnecessary-type-constraint
```

Please consider creating an isolated reproduction repo to make it easy for the volunteer maintainers debug your issue.
-->

```json
{
  "root": true,
  "parser": "@typescript-eslint/parser",
  "plugins": [
    "@typescript-eslint"
  ],
  "rules": {
    "@typescript-eslint/no-unnecessary-type-constraint": "error"
  }
}
```

**Expected Result**

Typescript eslint should not issue a warning as the suggested code (as below) 

**Actual Result**

The code provided in `working.tsx` is required due to how generics are parsed in Typescript/React, one cannot simply write `<T>`.

**Additional Info**
The attempted fix is invalid code:
```ts
export const GenericComponent = <T, = unknown>(props: GenericComponentProps<T>) => (
  <div>hello</div>
);
```

**Versions**
(Tested both)

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.2.0` |
| `@typescript-eslint/parser`        | `5.2.0` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `8.1.0` |
| `node`                             | `14.16.0` |

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.33.1-alpha.15` |
| `@typescript-eslint/parser`        | `4.33.1-alpha.15` |
| `TypeScript`                       | `4.4.4` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `14.16.0` |

**Comment by jacekkarczmarczyk** [comment]
not sure if related but here's similar case without using jsx:

```ts
interface Options<T> {
  success?: (trigger: HTMLElement, data: T) => void;
}
```

```
  41:36  error  Constraining the generic type `T` to `any` does nothing and is unnecessary  @typescript-eslint/no-unnecessary-type-constraint
```

**Comment by bradzacher** [comment]
We are a community run project. The **volunteer** maintainers spend most of their time triaging issues and reviewing PRs. This means that most issues *will not progress* unless a member of the community steps up and **champions** it.

If this issue is important to you - consider being that champion.
If not - please use the subscribe function and wait patiently for someone else to implement.

**Comment by graup** [comment]
If I am to propose a fix, how about one or some of these:

1. Find a good workaround and add that to the rule documentation. (Anyway `extends any` has always been a workaround for a JSX syntax limitation. So changing the workaround to a new workaround seems okay to me.)
2. Automatically ignore this rule in JSX files. (Is that possible?)
3. Remove this rule from `@typescript-eslint/recommended`. Or at least change it to a warning? That way, CI doesn't blow up as bad when updating @typescript-eslint.
3. Add a hint to the rule documentation that it's not compatible with JSX and should be disabled manually.

**Comment by bradzacher** [comment]
The fix is to update the rule to allow arrow functions to declare `extends unknown` if the generic also has a default set.

```ts
// Bad
function foo<T extends unknown>() {}
function bar<T extends unknown = unknown>() {}

// Bad
const x = function <T extends unknown>() {};
const y = function <T extends unknown = unknown>() {};

// Bad
const z = <T extends unknown>() => {};

// Good
const a = <T,>() => {};
const b = <T extends unknown = unknown>() => {};
```

**Comment by bradzacher** [comment]
https://github.com/microsoft/TypeScript/pull/47112
TS4.6 will include support for `const x = <T = unknown>() => {};`
So we don't really need to change the rule's logic

However - there is still a bug here because the fixer is incorrect.

**Comment by Josh-Cena** [comment]
The fixer also fixes

```ts
const GenericComponent = <T extends unknown, U extends unknown>(props: T) => {};
```

to

```ts
const GenericComponent = <T,, U,>(props: T) => {};
```

We should only add the trailing comma if there's only one generic argument _and_ there's no default

**Comment by graup** [comment]
One more case:

```ts
const GenericComponent = <T extends unknown,>(props: T) => {};
```
gets another comma too:
```ts
const GenericComponent = <T,,>(props: T) => {};
```

I'll make a PR with all of these test cases and hopefully a fix.

The thing is that if we produce `<T = string>` for `<T extends any = string>`, in JSX that is invalid syntax in TS < 4.6. Is it ok to produce invalid syntax in that case?

**Comment by Josh-Cena** [comment]
@graup Are you going to work on this recently?

**Comment by Josh-Cena** [comment]
@bradzacher Given #3509, if the plan is to eventually remove the fixer, do we need to fix the fixer here, or should we just remove it altogether? Not very familiar with ESLint's semver policy—is removing the fixer a BC?

**Comment by bradzacher** [comment]
@Josh-Cena no, removing a fixer is just a patch bump as it's a bug fix.

delete is probably the wrong word for the other issue.
It's true that we need to remove the autofixer because it can break the build. But there's no reason it could be a suggestion fixer instead. Suggestion fixers are allowed to break the build because they can only be applied manually via the IDE.

So we should still fix the logic because the logic can still be used.

**Comment by graup** [comment]
@Josh-Cena I'll get to this earliest in the first week of March. Feel free to take it. Here's how far I came so far: https://github.com/graup/typescript-eslint/compare/main..fix-4062-fixer (maybe you can reuse the tests, the first of which is still failing).
